### PR TITLE
Feat/describe

### LINKED ISSUE #215: 🤯 Cognitarium: Describe query implementation
## Purpose

Implement the `Describe` query message handling.

### PR DESCRIPTION
**Pull request by ccamel**
Closes #215 - providing the implementation for the `Describe` query message handling.

## Solution

The implementation is fairly simple, heavily relying on the Sparql engine developed by @amimart. The main idea is to transform a `Describe` request into a `Select` request.

**For named node:**

```sparql
describe <https://ontology.okp4.space/dataverse/dataspace/metadata/dcf48417-01c5-4b43-9bc7-49e54c028473>

=

select ?p ?o
where
  <https://ontology.okp4.space/dataverse/dataspace/metadata/dcf48417-01c5-4b43-9bc7-49e54c028473> ?p ?o
```

**For a variable:**

```sparql
prefix core: <https://ontology.okp4.space/core/>
describe ?x
where
  ?x core:hasDescription "A test Dataset."@en

=

select ?x ?x_p ?x_o
where
  ?x core:hasDescription "A test Dataset."@en
  ?x ?x_p ?x_o
```

## Breaking changes

I made an update to the contract to fix the resource type described. It was changed from VarOrNode to VarOrNamedNode because it doesn't make sense to describe a blank node.

## Details

The [oxigraph/rio](https://github.com/oxigraph/rio/) library used for serializing triplets requires string literals with lifetimes (`&str` - cf. [oxigraph/rio#triples](https://github.com/oxigraph/rio/blob/3bd01c2c977a0b01c918f6840cd05356477db358/api/src/model.rs#L270)). This requirement makes it challenging to directly use them as output from the `engine`. To address this, I introduced an intermediate DTO called atome that facilitates the transportation of strings before serialization.

## Improvements

A tweak is made in the code to handle the case of a resource designated by a named node differently from a resource designated by a variable, which is tedious. To do it properly, it would be advisable to introduce support for the [VALUES](https://www.w3.org/TR/sparql11-query/#inline-data) operator, which would allow binding a variable to its value. It would simplify the code on the one hand, and open up the possibilities on the other.

## Tests

I have added a number of tests, particularly to cover the features such as named nodes, variables, blank nodes, and multiple described resources. However, the error cases are not yet covered, which sadly affects the overall coverage.

### CODE REVIEW

**Review by amimart — CHANGES_REQUESTED** [review]
That's a nice piece thx! 🙏 

Just suggested a tiny improvement :)

File: contracts/okp4-cognitarium/src/contract.rs [hunk 1]
```diff
@@ -104,10 +115,122 @@ pub mod query {
...
+        if let VarOrNamedNode::NamedNode(iri) = &query.resource {
+            for b in &mut bindings {
+                b.insert(
+                    s.clone(),
+                    Value::URI {
+                        value: iri.to_owned(),
+                    },
+                );
+            }
+        }
```
[the remark this answers is not in the document]
Comment (ccamel) [reply contracts/okp4-cognitarium/src/contract.rs:205]: Ah ah, well spotted!

**Review by amimart — APPROVED** [review]

**Comment by ccamel** [comment]
Thanks @bdeneux for increasing test coverage :relaxed:

**Review by bdeneux — APPROVED** [review]
Nice ! That seems good :) 😁

**Pull request merged**